### ISSUE TITLE
Win64 : Error with non standard Chars with new versions

### ISSUE THREAD

**Issue by naeloob** [ncw/rclone#1212]
The new versions cannot find de local dir : 

-řá'ö

Full path : d:/test/-řá'ö

in 1.35 version works with it without problems.
I end with this dirname when i found errors with dirnames including these chars.

> What is your rclone version (eg output from `rclone -V`)
Right behaviour 1.35
Worng : Beta 1.35-149
> Which OS you are using and how many bits (eg Windows 7, 64 bit)
Win10 64Bits
> Which cloud storage system are you using? (eg Google Drive)
Gdrive
> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
Beta 1.35-149
d:\Soft\rclone-v1.35-windows-amd64>rclone-v1.35-149-g5fba913.exe copy d:/test RemoteCrypt:"/test" --max-age 2M -vvv
2017/03/05 16:13:21 DEBUG : --max-age 1440h0m0s to 2017-01-04 16:13:21.3337468 +0100 CET
2017/03/05 16:13:21 DEBUG : rclone: Version "v1.35-149-g5fba913β" starting with parameters ["rclone-v1.35-149-g5fba913.exe" "copy" "d:/test" "RemoteCrypt:/test" "--max-age" "2M" "-vvv"]
2017/03/05 16:13:24 INFO  : Encrypted Google drive root 'ot1ot51mktl1r8sfrthkudoik0': Modify window is 1ms
2017/03/05 16:13:24 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading ""
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading ""
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:25 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for checks to finish
2017/03/05 16:13:25 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for transfers to finish
2017/03/05 16:13:25 ERROR : Attempt 1/3 failed with 0 errors and: error reading source directory "-řá'ö": directory not found
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading ""
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading ""
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:25 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for checks to finish
2017/03/05 16:13:25 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for transfers to finish
2017/03/05 16:13:25 ERROR : Attempt 2/3 failed with 0 errors and: error reading source directory "-řá'ö": directory not found
2017/03/05 16:13:25 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading ""
2017/03/05 16:13:26 DEBUG : pacer: Rate limited, sleeping for 1.94777941s (1 consecutive low level retries)
2017/03/05 16:13:26 DEBUG : pacer: low level retry 1/10 (error googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded)
2017/03/05 16:13:26 DEBUG : pacer: Rate limited, sleeping for 2.082153551s (2 consecutive low level retries)
2017/03/05 16:13:26 DEBUG : pacer: low level retry 2/10 (error googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded)
2017/03/05 16:13:28 DEBUG : pacer: Resetting sleep to minimum 10ms on success
2017/03/05 16:13:28 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading ""
2017/03/05 16:13:28 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:30 DEBUG : Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:13:30 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for checks to finish
2017/03/05 16:13:30 INFO  : Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for transfers to finish
2017/03/05 16:13:30 ERROR : Attempt 3/3 failed with 0 errors and: error reading source directory "-řá'ö": directory not found
2017/03/05 16:13:30 Failed to copy: error reading source directory "-řá'ö": directory not found


1.35

d:\Soft\rclone-v1.35-windows-amd64>rclone.exe copy d:/test RemoteCrypt:"/test" --max-age 2M -vvv
2017/03/05 16:29:35 --max-age 1440h0m0s to 2017-01-04 16:29:35.60095 +0100 CET
2017/03/05 16:29:35 rclone: Version "v1.35" starting with parameters ["rclone.exe" "copy" "d:/test" "RemoteCrypt:/test" "--max-age" "2M" "-vvv"]
2017/03/05 16:29:35 dorukan: Saved new token in config file
2017/03/05 16:29:39 Encrypted Google drive root k1kj3idoe9apdjkfaghkab2go1': Modify window is 1ms
2017/03/05 16:29:39 Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading ""
2017/03/05 16:29:39 Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading ""
2017/03/05 16:29:39 Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:29:40 Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Finished reading "ysg36fw7a9cv02b3n45k1javd4/"
2017/03/05 16:29:40 Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for checks to finish
2017/03/05 16:29:40 -řá'ö/test.txt: Size and modification time the same (differ by -490.9µs, within tolerance 1ms)
2017/03/05 16:29:40 -řá'ö/test.txt: Unchanged skipping
2017/03/05 16:29:40 Encrypted Google drive root 'k1kj3idoe9apdjkfaghkab2go1': Waiting for transfers to finish
2017/03/05 16:29:40
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 1
Transferred:            0
Elapsed time:        4.4s
2017/03/05 16:29:40 Go routines at exit 5
```

**Comment by ncw** [comment]
What happens if you do

    rclone ls d:/test

What do you see there?

**Comment by naeloob** [comment]
d:\Soft\rclone-v1.35-windows-amd64>rclone ls d:/test
        4 -řá'ö/test.txt

d:\Soft\rclone-v1.35-windows-amd64>rclone-v1.35-149-g5fba913.exe ls d:/test
        4 -řá'ö/test.txt

Both versions show the same...

**Comment by naeloob** [comment]
Checked between 1.35 and "1.36" betas.
Latest working ok with that chars is rclone-v1.35-62-g9d331ceβ.
It fails from next one, rclone-v1.35-74-g48cdedcβ.

**Comment by ncw** [comment]
Excellent debugging - thanks.

What if you try with the `--old-sync` flag? I expect that will work.

**Comment by naeloob** [comment]
You're right!!
With --old-sync-method flag works.

Thank you for the support.

**Comment by ncw** [comment]
I'd like to find out why it isn't working... (I'm planning to remove the --old-sync flag eventually).

I've tried to replicate the problem but I haven't been able to.

How did you create that directory structure?  Could you make a bat script which creates the structure for me?

Is your disk in some kind of special encoding?  Or are some of the unicode characters not normalised properly?  That might explain it...

**Comment by naeloob** [comment]
I think it's just unicode.

I've uploaded a bat file that recreates the dir.
I've put jpg extension to be sure that it will be treated as a bin file.
Download and just change the extension jpg to bat.

![make](https://cloud.githubusercontent.com/assets/25527142/23653716/ca5ec19c-032d-11e7-88f6-d0dd997ba167.jpg)

**Comment by ncw** [comment]
I've managed to replicate the problem with the help of your .bat file - thanks!  I'll post more when I've worked out what is going on and have a beta for you to try!

**Comment by naeloob** [comment]
Thank you, great work!!

**Comment by ncw** [comment]
Looks like this **is** a unicode normalisation problem.

What is happening is that the directory objects (unlike the file objects) don't contain the un-normalised path.

This is going wrong because rclone now recurses through the directories in the sync code, rather than letting the remote do it.

There are two ways of fixing this

  1. store the un-normalised path in the directory object
    * change all the listing code to take a directory object (or be a directory object method)
  2. store a cache of un-normalised to normalised listings.
    * we do the equivalent of this using the dircache infrastructure
    * the other way would be a better way of doing this in general and would get rid of the need for dircache (mostly)

Either way that is quite an involved fix :-(

Is this something to do with the `chcp 65001` in your batch file.  What does that do?

This is also going to be a serious issue for macOS which loves un-normalised unicode.


```
C:\Users\Dev>x:\go\src\github.com\ncw\rclone\rclone.exe -vv copy t2 t3 --retries 1
2017/03/07 16:48:07 DEBUG : rclone: Version "v1.35-DEV" starting with parameters ["x:\\go\\src\\github.com\\ncw\\rclone\\rclone.exe" "-vv" "copy" "t2" "t3" "--retries" "1"]
2017/03/07 16:48:07 INFO  : Local file system at \\?\C:\Users\Dev\t3: Modify window is 100ns
2017/03/07 16:48:07 NOTICE: Local file system at \\?\C:\Users\Dev\t2: Cleaned "-r?'a´o¨" -> "-r'áö"  #### here is rclone normalising the path
2017/03/07 16:48:07 NOTICE: stat error is: GetFileAttributesEx \\?\C:\Users\Dev\t2\-r'áö: The system cannot find the file specified.
2017/03/07 16:48:07 INFO  : Local file system at \\?\C:\Users\Dev\t3: Waiting for checks to finish
2017/03/07 16:48:07 INFO  : Local file system at \\?\C:\Users\Dev\t3: Waiting for transfers to finish
2017/03/07 16:48:07 ERROR : Attempt 1/1 failed with 0 errors and: error reading source directory "-r'áö": directory not found
2017/03/07 16:48:07 Failed to copy: error reading source directory "-r'áö": directory not found

C:\Users\Dev>dir t2\-r'áö ##### normalised unicode here
 Volume in drive C has no label.
 Volume Serial Number is 5CD1-D250

 Directory of C:\Users\Dev\t2

File Not Found

C:\Users\Dev>dir t2
 Volume in drive C has no label.
 Volume Serial Number is 5CD1-D250

 Directory of C:\Users\Dev\t2

07/03/2017  15:47    <DIR>          .
07/03/2017  15:47    <DIR>          ..
07/03/2017  15:50    <DIR>          -r?'a´o¨  ##### un-normalised unicode here
               0 File(s)              0 bytes
               3 Dir(s)   6,261,989,376 bytes free

C:\Users\Dev>
```

**Comment by naeloob** [comment]
`chcp 65001` just changes the codepage of the command session to UTF-8.
https://ss64.com/nt/chcp.html

**Comment by ncw** [comment]
I've fixed this here

http://beta.rclone.org/v1.35-176-g5355881/ (uploaded in 15-30 mins)

Please re-open if you are still having problems!

**Comment by naeloob** [comment]
That's Great News!!! Thank You very Much. 
I will test and report.

Regards

**Comment by naeloob** [comment]
First Test seems to be OK!!!
I will do more, but i think it will work.

Thank you very much!!

**Comment by ncw** [comment]
thanks for testing :-)

**Comment by naeloob** [comment]
Sorry mate, but now i have the same problem with the same path on 1.36  linux64 version.
New sync method cannot find the local path, and with --old-sync-method works perfect.

Do you need more information about the issue?
Do you want me to open a new issue?

**Comment by ncw** [comment]
So it works with Windows now, but not with Linux is that correct?

As it is the same problem I'll open this issue again!

**Comment by ncw** [comment]
Can you so what you did before and make some shell commands I can run to reproduce?

**Comment by naeloob** [comment]
Yes, that seems to be the case.
It has any logic for you?

I will try to reproduce the creation of the dir...

**Comment by naeloob** [comment]
After several test, the problem seems to be that Local cannot find directories inside the Dir with non Standard Chars.

Ex : 
-řá'ö/file1
-řá'ö/samples/File2

So with new sync Method Local can read  -řá'ö/file1 but no -řá'ö/samples or inside it.
With --old-sync-method works.

It makes sense to you?


Tested with Linux.
I will test on Windows but i suposse it will be the same.

**Comment by ncw** [comment]
This should be fixed now for new uploads now that in #1477 we no longer normalize file names.
